### PR TITLE
Fix ST_Centroid and ST_Buffer for tiny geometries

### LINKED ISSUE #13194: NullPointerException in ST_Buffer
The following query fails with NPE in ST_Buffer:

```
> select st_buffer(st_buffer(st_point(177.50102959662, 64.726807421691), 0.0000000001), 0.00005)

java.lang.NullPointerException: input is null
        at java.base/java.util.Objects.requireNonNull(Objects.java:246)
        at com.facebook.presto.geospatial.serde.GeometrySerde.serialize(GeometrySerde.java:60)
        at com.facebook.presto.plugin.geospatial.GeoFunctions.stBuffer(GeoFunctions.java:367)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:707)
```

CC: @jagill

### PR DESCRIPTION
**Pull request by jagill**
Version 2.2.3 has two fixes of interest: 
- correct computation of centroids of very small polygons - https://github.com/Esri/geometry-api-java/pull/227
- correct computation of buffers around small polygons - https://github.com/Esri/geometry-api-java/pull/243

Fixes #13194 and #10629

```
== RELEASE NOTES ==

General Changes
-----------------
* Fix ST_Buffer for tiny geometries (#13194)
* Fix ST_Centroid for tiny geometries (#10629)
```

### CODE REVIEW

**Comment by jagill** [comment]
cc @mbasmanova

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 1]
```diff
@@ -195,6 +199,20 @@ public void testSTBuffer()
         assertInvalidFunction("ST_Buffer(ST_Point(0, 0), nan())", "distance is NaN");
     }
 
+    @Test
+    public void testSTBufferTinyPolygons()
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: I'd fold this test into `testSTBuffer`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 2]
```diff
@@ -207,16 +225,39 @@ public void testSTCentroid()
         assertCentroid("POLYGON ((1 1, 5 1, 3 4))", new Point(3, 2));
         assertCentroid("MULTIPOLYGON (((1 1, 1 3, 3 3, 3 1)), ((2 4, 2 6, 6 6, 6 4)))", new Point(3.3333333333333335, 4));
         assertCentroid("POLYGON ((0 0, 0 5, 5 5, 5 0, 0 0), (1 1, 1 2, 2 2, 2 1, 1 1))", new Point(2.5416666666666665, 2.5416666666666665));
+    }
 
+    @Test
+    public void testSTCentroidInvalidGeometry()
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: I'd fold this test into `testSTCentroid`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 3]
```diff
@@ -207,16 +225,39 @@ public void testSTCentroid()
...
+    {
         // invalid geometry
-        assertInvalidFunction("ST_Centroid(ST_GeometryFromText('MULTIPOLYGON (((4.903234300000006 52.08474289999999, 4.903234265193165 52.084742934806826, 4.903234299999999 52.08474289999999, 4.903234300000006 52.08474289999999)))'))", "Cannot compute centroid: .* Use ST_IsValid to confirm that input geometry is valid or compute centroid for a bounding box using ST_Envelope.");
+        assertApproximateCentroid("MULTIPOLYGON (((4.903234300000006 52.08474289999999, 4.903234265193165 52.084742934806826, 4.903234299999999 52.08474289999999, 4.903234300000006 52.08474289999999)))", new Point(4.9032343, 52.0847429), 1e-7);
+    }
+
+    @Test
+    public void testSTCentroidNumericStability()
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: I'd fold this test into `testSTCentroid`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 4]
```diff
@@ -207,16 +225,39 @@ public void testSTCentroid()
...
+    {
+        String wkt1 = "MULTIPOLYGON (((153.492818 -28.13729, 153.492821 -28.137291, 153.492816 -28.137289, 153.492818 -28.13729)))";
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: don't use a1, a2, a3... names for variables; here, I'd inline wkt1 and wkt2.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 5]
```diff
@@ -207,16 +225,39 @@ public void testSTCentroid()
...
     }
 
     private void assertCentroid(String wkt, Point centroid)
     {
         assertFunction(format("ST_AsText(ST_Centroid(ST_GeometryFromText('%s')))", wkt), VARCHAR, new OGCPoint(centroid, null).asText());
     }
 
+    private void assertApproximateCentroid(String wkt, Point centroid, double epsilon)
+    {
+        OGCGeometry geometry = OGCGeometry.fromText(wkt);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: I'd inline `geometry` variable

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 6]
```diff
@@ -207,16 +225,39 @@ public void testSTCentroid()
...
+        Slice inputSlice = GeometrySerde.serialize(geometry);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: I'd inline `inputSlice`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 7]
```diff
@@ -207,16 +225,39 @@ public void testSTCentroid()
...
+        OGCPoint calculatedCentroid = (OGCPoint) GeometrySerde.deserialize(stCentroid(inputSlice));
+        assertEquals(calculatedCentroid.X(), centroid.getX(), epsilon);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: I'd rename centroid to expectedCentroid and calculatedCentroid to centroid (or actualCentroid).

**Comment by jagill** [comment]
Addressed comments.  cc @mbasmanova

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 8]
```diff
@@ -193,6 +193,16 @@ public void testSTBuffer()
         // infinity() and nan() distance
         assertFunction("ST_AsText(ST_Buffer(ST_Point(0, 0), infinity()))", VARCHAR, "MULTIPOLYGON EMPTY");
         assertInvalidFunction("ST_Buffer(ST_Point(0, 0), nan())", "distance is NaN");
+
+        // For small polygons, there was a bug in ESRI that throw an NPE.  This
+        // was fixed (https://github.com/Esri/geometry-api-java/pull/243) to
+        // return an empty geometry instead. Ideally, these would return
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Would you add a note in ST_Buffer's documentation about this behavior?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 9]
```diff
@@ -384,33 +383,15 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
-                default:
-                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Unexpected geometry type: " + geometryType);
-            }
+            return serialize(geometry.centroid());
         }
         catch (RuntimeException e) {
             if (e instanceof PrestoException && ((PrestoException) e).getErrorCode() == INVALID_FUNCTION_ARGUMENT.toErrorCode()) {
                 throw e;
             }
             throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Cannot compute centroid: %s Use ST_IsValid to confirm that input geometry is valid or compute centroid for a bounding box using ST_Envelope.", e.getMessage()), e);
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @jagill Is this logic still needed? I see that you removed the test that triggered it.

Comment (jagill) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: I suppose not!

**Comment by jagill** [comment]
Comments address. cc @mbasmanova

**Pull request merged**